src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs

```diff
@@ -699,11 +660,50 @@ private static int NextSiblingIndex(in RenderTreeFrame frame, int frameIndex)
+                        // any parameter is of a type we don't know is immutable. In this case
+                        // we call SetParameters and it's up to the recipient to implement
+                        // whatever change-detection logic they want. Currently we only supply the new
+                        // set of parameters and assume the recipient has enough info to do whatever
+                        // comparisons it wants with the old values. Later we could choose to pass the
+                        // old parameter values if we wanted. By default, components always rerender
+                        // after any SetParameters call, which is safe but now always optimal for perf.
+
+                        // When performing hot reload, we want to force all components to re-render.
+                        // We do this using two mechanisms - we call SetParametersAsync even if the parameters
+                        // are unchanged and we ignore ComponentBase.ShouldRender.
+                        // Furthermore, when a hot reload edit removes component parameters, the component should be
+                        // disposed and reinstantiated. This allows the component's construction logic to correctly
+                        // re-initialize the removed parameter properties.
+
+                        var oldParameters = new ParameterView(ParameterViewLifetime.Unbound, oldTree, oldFrameIndex);
+                        var newParametersLifetime = new ParameterViewLifetime(diffContext.BatchBuilder);
+                        var newParameters = new ParameterView(newParametersLifetime, newTree, newFrameIndex);
+                        var isHotReload = HotReloadManager.Default.MetadataUpdateSupported && diffContext.Renderer.IsRenderingOnMetadataUpdate;
+
+                        if (isHotReload && newParameters.HasRemovedDirectParameters(oldParameters))
+                        {
+                            // Components with parameters removed during a hot reload edit should be disposed and reinstantiated
+                            RemoveOldFrame(ref diffContext, oldFrameIndex);
+                            InsertNewFrame(ref diffContext, newFrameIndex);
+                        }
+                        else
+                        {
+                            var componentState = oldFrame.ComponentStateField;
+
+                            // Preserve the actual componentInstance
+                            newFrame.ComponentStateField = componentState;
+                            newFrame.ComponentIdField = componentState.ComponentId;
+
+                            if (!newParameters.DefinitelyEquals(oldParameters) || isHotReload)
+                            {
+                                componentState.SetDirectParameters(newParameters);
+                            }
+
+                            diffContext.SiblingIndex++;
+                        }
```

Comment:
I'm curious about the reasoning for inlining all of this logic. Does it reduce evaluations meaningfully? If so then fine :)

The `AppendDiffEntriesForFramesWithSameSequence` method is pretty giant now so if the new logic could be extracted without cost that could help with readability, but not a big deal.

Reply:
This was something I debated with myself before settling on inlining the logic. My thought process was something like:

* We only want to increment `diffContext.SiblingIndex` if the component is not getting removed, and the logic determining whether that happens would be in the `UpdateRetainedChildComponent()` helper method.
* Moving `diffContext.SiblingIndex++` to the helper method makes it less obvious that it's happening at all, especially since it's incremented inline in the other `case`s
* We _could_ turn `UpdateRetainedChildComponent()` into `TryUpdateRetainedChildComponent()` and use the return value to determine whether to increment the sibling index inline, but this seems a bit overkill (and that pattern isn't used elsewhere in this file)
* The `case` for `RenderTreeframeType.Element` is already even longer than the new `case` for `RenderTreeFrameType.Component` (excluding the comment), so it might be best just to inline everything to avoid the problem of finely separating what stays inline and what doesn't

Definitely open to having my mind changed, though!

Reply:
OK, sounds totally fine. It's clear you've thought it through well so let's go with what you're saying!